bazel_repositories_priv

```diff
@@ -0,0 +1,292 @@
+#!/usr/bin/python3
+"""\
+Privileged part of OBS Source Service to download dependencies of projects
+using Bazel. It uses administrator rights to manage network namespaces.
+"""
+
+import argparse
+import ctypes
+import glob
+import hashlib
+import json
+import logging
+import multiprocessing
+from multiprocessing import connection as m_connection
+import os
+import pathlib
+import pwd
+import re
+import shutil
+import subprocess
+import tarfile
+import typing
+from urllib import error as urllib_error
+from urllib import parse as urllib_parse
+from urllib import request as urllib_request
+import uuid
+
+
+app_name = "obs-service-bazel_repositories"
+description = __doc__
+
+logging.basicConfig(level=logging.INFO)
+log = logging.getLogger(app_name)
+
+libc = ctypes.cdll.LoadLibrary('libc.so.6')
+
+CLONE_NEWUSER = 0x10000000
+CLONE_NEWNET = 0x40000000
+
+ROOT_CACHEDIR = "BAZEL_CACHE"
+CACHEDIR = "BAZEL_CACHE/content_addressable/sha256"
+
+user = None
+
+
+def as_user():
+    """Change user from root to the unprivileged user. Used in functions which
+    run in forked processes.
+    """
+    log.debug(f"logging as user {user}")
+    user_passwd = pwd.getpwnam(user)
+    os.setgroups([user_passwd.pw_gid])
+    os.setgid(user_passwd.pw_gid)
+    os.setuid(user_passwd.pw_uid)
+    os.environ["HOME"] = f"/home/{user}"
+    os.environ["USER"] = user
+
+
+def find_rootdir(filename_wildcard: str) -> str:
+    """Return the main directory inside the tarball with sources of packaged
+    project.
+    """
+    root_dir = None
+    filename = glob.glob(filename_wildcard)[0]
+    tar = tarfile.open(filename)
+    for tarinfo in tar:
+        root_dir = tarinfo.name
+        break
+    tar.close()
+    return root_dir
+
+
+def bazel_clean(root_dir: str):
+    """Clean Bazel cache.
+    """
+    os.chdir(root_dir)
+    as_user()
+    subprocess.run(["bazel", "clean", "--expunge"], check=True)
+
+
+def sub(conn: m_connection.Connection, netns_name: str, root_dir: str,
+        bazel_args: str, target: str):
+    """The main sub process function which runs Bazel inside a network
+    namespace without Internet connectivity.
+    """
+    log.debug(f"sub target: {target}")
+    os.chdir(root_dir)
+
+    nsfd = os.open(os.path.join("/var/run/netns", netns_name), os.O_RDONLY)
+    libc.setns(nsfd, CLONE_NEWNET)
+
+    as_user()
+
+    args = ["bazel", "fetch", "--repository_cache=../BAZEL_CACHE"]
+    if bazel_args:
+        args += bazel_args.split()
+    args += [target]
+
+    log.debug(f"args: {args}")
+
+    p = subprocess.Popen(args, stdout=subprocess.PIPE,
+                         stderr=subprocess.STDOUT,
+                         universal_newlines=True)
+    stdout = ""
+    for line in iter(p.stdout.readline, ""):
+        line_s = line.strip()
+        log.debug(f"bazel fetch: {line_s}")
+        stdout += line
+    p.stdout.close()
+    returncode = p.wait()
+
+    urls = re.findall(r"https?://(?:[a-zA-Z]|[0-9]|[-_@.&/+])+", stdout)
```

Comment:
I'm not sure about this regex.  Splitting the three character classes into three separate or'd expressions is going to slow down parsing a little, I think.  You should be able to group those into one big class, and then you don't need the outer `(?:)` to apply the `+`.  The 0-9/a-z/A-Z and underscore also can be collapsed into just `\W` by setting re.ASCII (or embedding `(?a)` in the expression).  For example:
```suggestion
    urls = re.findall(r"(?a)https?://[-_@.&/+\W]+", stdout)
```
However, even easier would be to split the output up on whitespace and use `urlparse.urlparse(token)` in a try/except block where you `continue` the loop on exception or add to the list if no exception.  Then there's no regex to worry about (let the urlparse module deal with it), and you have pretty valid URLs.  Maybe a pre-check to see if the token `.startswith('http')`. :shrug:

Reply:
Th idea with spliting and using urlparse sounds good, I will try it. Especially because I can do it in the same loop where I print each line of stdout.

Reply:
Nah, I can#t use `urlparse` so easy here. Usually, the error from Bazel looks like:

```
java.io.IOException: Error downloading [https://github.com/lightstep/lightstep-tracer-cpp/archive/3efe2372ee3d7c2138d6b26e542d757494a7938d.tar.gz] to /home/vadorovsky/.cache/bazel/_bazel_vadorovsky/27f2b801d540e1f6e0caa939df2fd2cd/external/com_lightstep_tracer_cpp/3efe2372ee3d7c2138d6b26e542d757494a7938d.tar.gz: Network is unreachable (connect failed)
```

Using just `split` and `urlparse` is not enough:

```
In [5]: line = "java.io.IOException: Error downloading [https://github.com/lightstep/lightstep-tracer-cpp/archive/3efe2372ee3d7c2138d6b26e542d757494a7938d.tar.gz] to /home/vadorovsky/.cache/bazel/_bazel_vadorovsky/27f2b801d540e1f6e0caa939df2fd2cd/external/com_lightstep_trac
   ...: er_cpp/3efe2372ee3d7c2138d6b26e542d757494a7938d.tar.gz: Network is unreachable (connect failed)"                                                                                                                                                                          

In [6]: for token in line.split(): 
   ...:     print(urllib_parse.urlparse(token)) 
   ...:                                                                                                                                                                                                                                                                           
ParseResult(scheme='java.io.ioexception', netloc='', path='', params='', query='', fragment='')
ParseResult(scheme='', netloc='', path='Error', params='', query='', fragment='')
ParseResult(scheme='', netloc='', path='downloading', params='', query='', fragment='')
ParseResult(scheme='', netloc='', path='[https://github.com/lightstep/lightstep-tracer-cpp/archive/3efe2372ee3d7c2138d6b26e542d757494a7938d.tar.gz]', params='', query='', fragment='')
ParseResult(scheme='', netloc='', path='to', params='', query='', fragment='')
ParseResult(scheme='', netloc='', path='/home/vadorovsky/.cache/bazel/_bazel_vadorovsky/27f2b801d540e1f6e0caa939df2fd2cd/external/com_lightstep_tracer_cpp/3efe2372ee3d7c2138d6b26e542d757494a7938d.tar.gz:', params='', query='', fragment='')
ParseResult(scheme='', netloc='', path='Network', params='', query='', fragment='')
ParseResult(scheme='', netloc='', path='is', params='', query='', fragment='')
ParseResult(scheme='', netloc='', path='unreachable', params='', query='', fragment='')
ParseResult(scheme='', netloc='', path='(connect', params='', query='', fragment='')
ParseResult(scheme='', netloc='', path='failed)', params='', query='', fragment='')
```

Of course I could clean splitted tokens from brackets, but it's much more effort than using just a regex.

I will use your `urls = re.findall(r"(?a)https?://[-_@.&/+\W]+", stdout)` pattern. :)

Reply:
Darn; that sounded like such a great idea in my head. :D

It's a shame python doesn't have perl's Regexp::Common::URI module. https://metacpan.org/pod/Regexp::Common::URI::http